### ISSUE TITLE
Persistent Cache with cache manager level serializer

### ISSUE THREAD

**Issue by JSAhuja** [ehcache/ehcache3#617]
Persistent Cache is not picking up the serializer defined at the Cache Manager level.
````
DefaultSerializationProviderConfiguration serializer =    new DefaultSerializationProviderConfiguration().addSerializerFor(java.lang.Long .class, CustomSerializerLong.class);
PersistentCacheManager persistentCacheManager = CacheManagerBuilder.newCacheManagerBuilder()
        .with(new CacheManagerPersistenceConfiguration(new File(getClass().getClassLoader().getResource(".").toURI().getPath(), "myData")))
        .withCache("persistent-cache", CacheConfigurationBuilder.newCacheConfigurationBuilder()
            .withResourcePools(ResourcePoolsBuilder.newResourcePoolsBuilder()
                .heap(10, EntryUnit.ENTRIES)
                .disk(10L, MemoryUnit.MB, true))
            .buildConfig(Long.class, String.class))
        .using(serializer)
        .build(true);
    Cache<Long, String> cache = persistentCacheManager.getCache("persistent-cache", Long.class, String.class);
    persistentCacheManager.close();
````
Console log for the serializer registered
````
15:33:54.578 [main] INFO  org.ehcache.EhcacheManager - Cache 'persistent-cache' is getting created in EhcacheManager.
15:33:54.583 [main] INFO  o.e.i.p.DefaultLocalPersistenceService - Reusing C:\Users\jia\Documents\SourceTree\ehcachetest\copier\target\test-classes\myData\persistent-cache_4773b8fc9608d7ffb5160821d40afbfe5cec777f
15:33:54.589 [main] INFO  o.e.i.p.DefaultLocalPersistenceService - Reusing C:\Users\jia\Documents\SourceTree\ehcachetest\copier\target\test-classes\myData\persistent-cache_4773b8fc9608d7ffb5160821d40afbfe5cec777f\DefaultSerializationProvider-Key
15:33:54.612 [main] INFO  o.e.s.s.DefaultSerializationProvider - Serializer for <java.lang.Long> : org.ehcache.internal.serialization.CompactPersistentJavaSerializer@3252ac20
15:33:54.616 [main] INFO  o.e.i.p.DefaultLocalPersistenceService - Reusing C:\Users\jia\Documents\SourceTree\ehcachetest\copier\target\test-classes\myData\persistent-cache_4773b8fc9608d7ffb5160821d40afbfe5cec777f\DefaultSerializationProvider-Value
15:33:54.618 [main] INFO  o.e.s.s.DefaultSerializationProvider - Serializer for <java.lang.String> : org.ehcache.internal.serialization.CompactPersistentJavaSerializer@1b1dafa4
15:33:54.624 [main] INFO  o.e.spi.copy.DefaultCopyProvider - Copier for <java.lang.Long> : org.ehcache.internal.copy.IdentityCopier@6079bbda
15:33:54.625 [main] INFO  o.e.spi.copy.DefaultCopyProvider - Copier for <java.lang.String> : org.ehcache.internal.copy.IdentityCopier@581c16b1
15:33:54.699 [main] INFO  o.e.i.p.DefaultLocalPersistenceService - Reusing C:\Users\jia\Documents\SourceTree\ehcachetest\copier\target\test-classes\myData\persistent-cache_4773b8fc9608d7ffb5160821d40afbfe5cec777f\offheap-disk-store
15:33:54.833 [main] INFO  o.e.i.store.disk.OffHeapDiskStore - The index for data file ehcache-disk-store.data is more recent than the data file itself by 64ms : this is harmless.
15:33:54.925 [main] INFO  c.ehcache.Ehcache-persistent-cache - Initialize successful.
15:33:54.925 [main] INFO  org.ehcache.EhcacheManager - Cache 'persistent-cache' created in EhcacheManager.
15:33:54.925 [main] INFO  org.ehcache.EhcacheManager - Initialize successful.
````

It is to be noted that if persistent & Disk is not configured the same thing works.

**Comment by ljacomet** [comment]
@JSAhuja This is not a reproducible test case, one critical element is missing: the custom serializer!

**Comment by JSAhuja** [comment]
Attaching the serializer:
````
@Serializer.Transient
public class CustomSerializerLong implements Serializer<Long> {
  private static final Charset CHARSET = Charset.forName("US-ASCII");

  public CustomSerializerLong(ClassLoader classLoader) {
  }

  @Override
  public ByteBuffer serialize(Long object) {
    ByteBuffer byteBuffer = ByteBuffer.allocate(8);
    byteBuffer.putLong(object).flip();
    return byteBuffer;
  }

  @Override
  public Long read(ByteBuffer binary) throws ClassNotFoundException {
    long l = binary.getLong();
    return l;
  }

  @Override
  public boolean equals(Long object, ByteBuffer binary) throws ClassNotFoundException {
    return object.equals(read(binary));
  }

  @Override
  public void close() {
  }
}
````

**Comment by JSAhuja** [comment]
The custom serializer I am using is not of type persitent. If I mark it as persistent then it gets picked up.
But the same serilizer without the persistent annotation works fine when defined at the cache level.

**Comment by albinsuresh** [comment]
There is a serious inconsistency in terms of the usage of serializer annotations between the cache level and cache manager level configurations. The cache manager level serializer configurations are subjected to more checks than a cache level one. For example, you can register a serializer that is unannotated at the cache level. But if you try to register the same at the cache manager level it would throw. I believe we need to replicate these checks in the cache level configuration as well for a consistent behaviour.

**Comment by alexsnaps** [comment]
Couple of remarks here:
  - The `Serializer` _is_ annotated `@Serializer.Transient`?
 - Do we really need both annotations? If `Persistent` implies not `Transient` and vice-verso, why can't the absence of the annotation mean the "opposite" ?
 - Why is this even an annotation? Seems like an abuse of such... Or again certainly that _either or_ is required... Use the type-system for these things? 

Wonder whether we shouldn't step back here for a second. That being said, I realize I wasn't part of any of these discussion and there maybe good reason for the choices made... But all in all there is some confusion in the net result of those reasons then...

**Comment by alexsnaps** [comment]
Following some discussions with @chrisdennis @ljacomet @albinsuresh we have decided to get rid of the both annotations. They also force the user to have one (`ClassLoader`) and/or two (`ClassLoader`, `FileBasedPersistenceContext`) arg constructors to match the annotation(s). It's just additional burden put on the user. 

Instead of looking for the annotations, we should directly look for the constructors. One additional change we have decided on, is to force having a two arg constructor even if your `Serializer` implementation doesn't require the `FileBasedPersistenceContext` when used with persistent `Cache`. It's easy enough for the user to call :

```java
MySerializer implements Serializer {

  // Registers in the DefaultSerializationProvider#transientProvider
  MySerializer(ClassLoader cl) { 
    this.classLoader = cl;
  } 

  // Registers in the DefaultSerializationProvider#persistentProvider
  MySerializer(ClassLoader cl, FileBasedPersistenceContext pc) {
    this(cl);
  }
}
```

But avoids bad surprises, as we then are able to fail fast in case the expected constructor isn't present.
